### PR TITLE
fix: streamline publish function to prevent loss of links

### PR DESCRIPTION
**Pull request by Bento007**
- #TICKET_NUMBER

### Reviewers
**Functional:** 

**Readability:** 

---

## Changes
- Reorganize publish logic.
- Add __repr__ to make debugging easier
- Add additional publish revision test with refreshed dataset and new links.
- don't change published_at if tombstoned

### CODE REVIEW

File: backend/corpora/common/entities/collection.py [hunk 1]
```diff
@@ -245,23 +244,49 @@ def publish(self, data_submission_policy_version):
...
-                **revision,
-            )
+                **revision, commit=False
+            )  # This function call deletes old links (keep_links param defaults to False)
             for link in self.links:
-                CollectionLink(link).update(collection_id=self.revision_of, commit=False)
-            is_existing_collection = True
-
+                link.collection_id = self.revision_of
+            self.session.flush()
```
Comment (danieljhegeman) [inline backend/corpora/common/entities/collection.py:287]: @Bento007 I just added this line. I am pretty sure this is necessary to make 100% sure that the links are moved to the published object rather than themselves deleted. I think we had an inverse of this situation with the original bug which meant that the code usually produced the bug but occasionally worked (as per Jason's experience); without this line, I think the code would usually work but could possibly produce the bug.

I think the combination of reassigning the links and then the subsequent `delete` call does not have deterministic behavior with SQLAlchemy.

Basically, the flush is required to make 100% sure that the db transaction first sees that the revision links now belong to the published collection _before_ the revision object is deleted (in which case the links would themselves be deleted).

**Review by danieljhegeman — APPROVED** [review]

File: backend/corpora/common/entities/collection.py [hunk 2]
```diff
@@ -245,23 +244,49 @@ def publish(self, data_submission_policy_version):
...
+                if original and public_collection.check_has_dataset(original):
+                    dataset_is_changed = original.publish_revision(revised_dataset, now)
+                    if dataset_is_changed:
+                        has_dataset_changes = True
+                else:
+                    # The dataset is new
+                    revised_dataset.publish_new(now)
+                    has_dataset_changes = True
+            self.session.flush()
+            # calling expire because the above code updates and deletes rows related to this object. If we don't expire
```
Comment (Bento007) [inline backend/corpora/common/entities/collection.py:265]: @danieljhegeman
